SpatialGDK/Source/SpatialGDK/Private/Interop/RPCs/RPCService.cpp

```diff
@@ -0,0 +1,339 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "Interop/RPCs/RPCService.h"
+
+namespace SpatialGDK
+{
+RPCService::RPCService(const FSubView& InRemoteSubView, const FSubView& InLocalAuthSubView)
+	: RemoteSubView(&InRemoteSubView)
+	, LocalAuthSubView(&InLocalAuthSubView)
+{
+}
+
+void RPCService::AddRPCQueue(FName QueueName, RPCQueueDescription&& Desc)
+{
+	Queues.Add(QueueName, MoveTemp(Desc));
+}
+
+void RPCService::AddRPCReceiver(FName ReceiverName, RPCReceiverDescription&& Desc)
+{
+	Receivers.Add(ReceiverName, MoveTemp(Desc));
+}
+
+void RPCService::AdvanceSenderQueues()
+{
+	for (const EntityDelta& Delta : LocalAuthSubView->GetViewDelta().EntityDeltas)
+	{
+		switch (Delta.Type)
+		{
+		case EntityDelta::UPDATE:
+		{
+			for (const ComponentChange& Change : Delta.ComponentUpdates)
```

Comment:
I thought this was a two pass  update ComponentUpdates at first, maybe this could be clearer by wrapping this loop into a lambda / free function? Something like:
```
auto SendUpdate = [](const ComponentSpan<ComponentChange>& Changes, bool AddUpdateCtx) {
	  for (const ComponentChange& Change : Changes)
	  {
		  RPCReadingContext Ctx;
		  Ctx.EntityId = Delta.EntityId;
		  Ctx.ComponentId = Change.ComponentId;
		  if (AddUpdateCtx) Ctx.Update = Change.Update;
		  Ctx.Fields = Schema_GetComponentDataFields(Change.CompleteUpdate.Data);
	  
		  for (auto& QueueEntry : Queues)
		  {
			  if (QueueEntry.Value.Sender->GetComponentsToReadOnUpdate().Contains(Ctx.ComponentId))
			  {
				  QueueEntry.Value.Sender->OnUpdate(Ctx);
			  }
		  }
	  }
} ;

SendUpdate(Delta.ComponentUpdates, /*AddUpdateCtx*/ true);
SendUpdate(Delta.ComponentsRefreshed, /*AddUpdateCtx*/ false);
```

Reply:
Change.CompleteUpdate and Change.Data have different types.
It's not clear to me that putting these two loops through common code would result in cleaner code.
I did something like that in SpatialRoutingSystem::ProcessUpdate, but the setup is yet again a bit different.

Reply:
You could maybe consider creating two different constructors for the ctx object. The issues I have with this amount of repeated code that it can be difficult to spot the differences.